pennylane/pulse/parametrized_hamiltonian_pytree.py

```diff
@@ -0,0 +1,135 @@
+from dataclasses import dataclass
+from typing import Callable, Optional, Tuple, Union
+
+import jax
+import jax.numpy as jnp
+from jax.experimental import sparse
+from jax.tree_util import register_pytree_node_class
+
+import pennylane as qml
+
+from .parametrized_hamiltonian import ParametrizedHamiltonian
+
+
+@register_pytree_node_class
+@dataclass
+class ParametrizedHamiltonianPytree:
+    """Jax pytree class that represents a ``ParametrizedHamiltonian``."""
+
+    mat_fixed: Optional[Union[jnp.ndarray, sparse.CSR]]
+    mats_parametrized: Tuple[Union[jnp.ndarray, sparse.CSR], ...]
+    coeffs_parametrized: Tuple[Callable]
+
+    @staticmethod
+    def from_hamiltonian(H: ParametrizedHamiltonian, dense: bool = False, wire_order=None):
+        """Convert a ``ParametrizedHamiltonian`` into a jax pytree object.
+
+        Args:
+            H (ParametrizedHamiltonian): parametrized Hamiltonian to convert
+            dense (bool, optional): Decide wether a dense/sparse matrix is used. Defaults to False.
+            wire_order (list, optional): Wire order of the returned ``JaxParametrizedOperator``.
+                Defaults to None.
+
+        Returns:
+            ParametrizedHamiltonianPytree: pytree object
+        """
+        make_array = jnp.array if dense else sparse.CSR.fromdense
+
+        if len(H.ops_fixed) > 0:
+            mat_fixed = make_array(qml.matrix(H.H_fixed(), wire_order=wire_order))
+        else:
+            mat_fixed = None
+
+        mats_parametrized = tuple(
+            make_array(qml.matrix(op, wire_order=wire_order)) for op in H.ops_parametrized
+        )
+        return ParametrizedHamiltonianPytree(mat_fixed, mats_parametrized, H.coeffs_parametrized)
+
+    def __call__(self, pars, t):
+        if self.mat_fixed is not None:
+            ops = (self.mat_fixed,)
+            coeffs = (1,)
+        else:
+            ops = ()
+            coeffs = ()
+        coeffs = coeffs + tuple(f(p, t) for f, p in zip(self.coeffs_parametrized, pars))
+        return LazyDotPytree(coeffs, ops + self.mats_parametrized)
+
+    def tree_flatten(self):
+        """Function used by ``jax`` to flatten the JaxParametrizedOperator.
+
+        Returns:
+            tuple: tuple containing the matrices and the parametrized coefficients defining the class
+        """
+        matrices = (self.mat_fixed, self.mats_parametrized)
+        param_coeffs = self.coeffs_parametrized
+        return (matrices, param_coeffs)
+
+    @classmethod
+    def tree_unflatten(cls, param_coeffs: tuple, matrices: tuple):
+        """Function used by ``jax`` to unflatten the JaxParametrizedOperator.
+
+        Args:
+            param_coeffs (tuple): tuple containing the parametrized coefficients of the class
+            matrices (tuple): tuple containing the matrices of the class
+
+        Returns:
+            JaxParametrizedOperator: a JaxParametrizedOperator instance
+        """
+        return cls(*matrices, param_coeffs)
```

Comment:
Out of curiosity, what is this function doing? We have three attributes `mats_fixed`, `mats_parametrized`, and `coeff_parametrized`. From the name I would expect this method to pack them in their tree structure (which is what exactly?). 

Also: How come this choice of signature of the method?

Reply:
Are you asking about the `cls(*matrices, param_coeffs)`? `cls` here corresponds to `ParametrizedHamiltonianPytree`. This line is creating an instance of this class.

Reply:
I meant the whole method `pytree_unflatten`